ops/model.py

```diff
@@ -26,20 +26,85 @@
 import time
 import typing
 import weakref
-from abc import ABC, abstractmethod
-from collections.abc import Mapping, MutableMapping
+from abc import ABC, ABCMeta, abstractmethod
 from pathlib import Path
 from subprocess import PIPE, CalledProcessError, run
+from typing import (
+    Any,
+    BinaryIO,
+    Dict,
+    Iterable,
+    List,
+    Mapping,
+    MutableMapping,
+    Optional,
+    Sequence,
+    Set,
+    TextIO,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
 
 import ops
 import ops.pebble as pebble
 from ops._private import yaml
 from ops.jujuversion import JujuVersion
 
 if typing.TYPE_CHECKING:
-    _RelationMappingType = typing.Mapping[str, typing.List['Relation']]
-    _StorageMappingType = typing.Mapping[str, typing.List['Storage']]
-    _BindingMappingType = typing.Mapping[str, 'Binding']
+    from typing_extensions import TypedDict
+
+    _StorageDictType = Dict[str, Optional[List['Storage']]]
+    _BindingDictType = Dict[Union[str, 'Relation'], 'Binding']
+    Numerical = Union[int, float]
+
+    # all types that can be (de) serialized to json(/yaml) fom Python builtins
+    JsonObject = Union[Numerical, bool, str,
+                       Dict[str, 'JsonObject'],
+                       List['JsonObject'],
+                       Tuple['JsonObject', ...]]
+
+    # a k8s spec is a mapping from names/"types" to json/yaml spec objects
+    _K8sSpec = Mapping[str, JsonObject]
+
+    _StatusDict = TypedDict('_StatusDict', {'status': str, 'message': str})
+
+    # the data structure we can use to initialize pebble layers with.
+    # todo: replace with pebble._LayerDict (a TypedDict) when pebble.py is typed
+    _LayerDict = Dict[str, '_LayerDict']
+    _Layer = Union[str, _LayerDict, pebble.Layer]
+
+    # mapping from relation name to a list of relations
+    _RawRelationMapping = Dict[str, Optional[List['Relation']]]
+    # mapping from relation name to relation metadata
+    _RelationsMeta = Dict[str, ops.charm.RelationMeta]
+    # mapping from container name to container metadata
+    _ContainerMeta = Dict[str, ops.charm.ContainerMeta]
```

Comment:
but these types are not a mapping from string to FooMeta yaml objects, they are a mapping from string to actual FooMeta instances. 

I think `RelationNameToRelationMeta` is ugly though.

`FooMetas`?

Reply:
made them all _Raw for now, but I'm not convinced. Either way they're not user-facing types, so it's enough if we know what they are.

Reply:
Hmm... Yeah I could maybe go for `_FooMetas`.

Reply:
I would keep _FoosMeta_Raw because `relations_meta: '_RelationsMeta_Raw'` looks less ugly than `relations_meta: '_RelationsMetas'` or `relations_meta: '_RelationMetas'`.